### DISCUSSION TITLE
Is there a way to create custom components that support the `render` prop?

### DISCUSSION THREAD

**Discussion by gnapse** [ariakit/ariakit#3578] [Questions] [answered]
I've dug into the code, and realized that the `render` prop used for composition is powered by [Ariakit's internal `createElement` function](https://github.com/ariakit/ariakit/blob/ec6a39ea205e135b96784bf0bd09b053737806ba/packages/ariakit-react-core/src/utils/system.tsx#L34).

Is there a way to leverage this function to create components outside Ariakit that also support [Ariakit's composition pattern](https://ariakit.org/guide/composition)?

Bonus question: what's [the `wrapElement` prop](https://github.com/ariakit/ariakit/blob/ec6a39ea205e135b96784bf0bd09b053737806ba/packages/ariakit-react-core/src/utils/system.tsx#L52-L54)? Is it documented somewhere?

**Comment by diegohaz** [comment, marked as answer]
Yes, you can use the [`Role`](https://ariakit.org/reference/role) component:

```jsx
function MyComponent(props) {
  return <Role.div render={props.render} />;
}
```

`wrapElement` is a lower-level prop used by internal hooks to wrap the underlying element with other components like providers and React portal. You likely will never need it unless you're building custom component hooks like `useDialog`, `usePopover`, etc. Since those hooks don't return the JSX elements, but just the props to be passed to other elements, they use this custom `wrapElement` prop that will be later consumed by Ariakit's own `createElement` or `Role`.
